src/ql/pass/ana/visualize/detail/common.cc

```diff
@@ -18,85 +19,95 @@ namespace detail {
 
 using namespace utils;
 
-Vec<GateProperties> parseGates(const ir::compat::ProgramRef &program) {
-    Vec<GateProperties> gates;
+class GateCollector : public ir::RecursiveVisitor {
```

Comment:
There is a function in src/ql/ir/consistency.cc that is called "check_cycles_non_decreasing" and that checks that all instructions in a new-IR circuit have non-decreasing cycle numbers.

The invariant is also described in ir.tree:
`a program that has not yet been scheduled can be represented with any non-decreasing cycle assignment`

I looked at the reading pass and it does "naively schedule" all instructions in consecutive cycles.
An interesting point is that should this "naive scheduling" actually care about instruction duration? To me this in `read.cc` is not proper:
```
                    // If scheduling information is discarded, increment the
                    // cycle number at the end of each instruction.
                    if (options.schedule_mode != ScheduleMode::KEEP) {
                        cycle++;
                    }
```

as well as this:
```
            // The cycle counter increments at the end of each bundle if
            // scheduling information is retained. Otherwise it is incremented
            // at the end of each instruction.
            if (options.schedule_mode == ScheduleMode::KEEP) {
                if (!ql_bundle.empty() || !found_pragma) {
                    cycle++;
                }
            }
```

I think cycle should be incremented by the duration of the instruction. When there is a bundle, cycle should be incremented by the max duration of the contained instruction. But that's a different change.

To come back to "a new IR circuit is scheduled always", is there no value in displaying a non-scheduled circuit? Considering the invariant.